### PR TITLE
Lcra/refactor workspace openapi1 prod 14206

### PR DESCRIPTION
**Pull request by Leopold-Cramer**

### CODE REVIEW

File: workspace/src/main/openapi/workspace.yaml [hunk 1]
```diff
@@ -797,11 +748,146 @@ components:
...
+        sendScenarioRunToEventHub:
+          type: boolean
+          description: default setting for all Scenarios and Run Templates
+        sendScenarioMetadataToEventHub:
+          type: boolean
+          description: Set this property to false to not send scenario metadata
+        datasetCopy:
+          type: boolean
+          description: Activate the copy of dataset on scenario creation
+      required:
```
Comment (sjoubert) [inline workspace/src/main/openapi/workspace.yaml:864]: Why are any of the fields required for the update request? Why do I have to provide the existing key/name if I only want to change the tags for example?

File: workspace/src/main/openapi/workspace.yaml [hunk 2]
```diff
@@ -882,34 +970,49 @@ components:
...
           defaultRunTemplateDataset:
             run1: "data1"
-        sendInputToDataWarehouse: null
-        useDedicatekdEventHubNamespace: false
-        dedicatedEventHubSasKeyName: null
-        dedicatedEventHubAuthenticationStrategy: null
+        sendInputToDataWarehouse: false
+        useDedicatedEventHubNamespace: false
+        dedicatedEventHubSasKeyName: "RootManageSharedAccessKey"
+        dedicatedEventHubAuthenticationStrategy: "SHARED_ACCESS_POLICY"
```
Comment (sjoubert) [inline workspace/src/main/openapi/workspace.yaml:1015]: Did you do an analysis of the fields we can remove or is it scheduled for later? For example I don't think these fields (and others in the Workspace object) are used/useful anymore with the new orchestrator. There is a huge cleanup to do on Workspace in particular if I remember.

**Review by sjoubert — CHANGES_REQUESTED** [review]

**Review by jreynard-code — CHANGES_REQUESTED** [review]
Some remarks ;)
GJ

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 3]
```diff
@@ -106,56 +109,72 @@ internal class WorkspaceServiceImpl(
     return result
   }
 
-  override fun findWorkspaceById(organizationId: String, workspaceId: String): Workspace {
+  override fun getWorkspace(organizationId: String, workspaceId: String): Workspace {
     return updateSecurityVisibility(getVerifiedWorkspace(organizationId, workspaceId))
   }
 
-  override fun createWorkspace(organizationId: String, workspace: Workspace): Workspace {
+  override fun createWorkspace(organizationId: String, workspaceCreateRequest: WorkspaceCreateRequest): Workspace {
```
Comment (jreynard-code) [inline workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:116]: This method is broken. 
E.g: If I create a workspace and I specified a description in WorkspaceCreateRequest, its not saved.

Could you please check that and add tests on CRUD Workspace methods please ?

**Review by jreynard-code — CHANGES_REQUESTED** [review]
Some buggy changes.
Go deeper in changes please and add some tests to be sure that it works as it should ;)

File: run/src/integrationTest/kotlin/com/cosmotech/run/service/RunServiceIntegrationTest.kt [hunk 4]
```diff
@@ -150,20 +151,20 @@ class RunServiceIntegrationTest : CsmRunTestBase() {
...
 
-    workspace = mockWorkspace(organizationSaved.id, solutionSaved.id!!, "Workspace")
+    workspace = makeWorkspaceCreateRequest(organizationSaved.id, solutionSaved.id!!, "Workspace")
     workspaceSaved = workspaceApiService.createWorkspace(organizationSaved.id, workspace)
 
     runnerSaved =
         mockRunner(
             organizationSaved.id,
-            workspaceSaved.id!!,
+          workspaceSaved.id,
```
[the remark this answers is not in the document]
Comment (Leopold-Cramer) [reply run/src/integrationTest/kotlin/com/cosmotech/run/service/RunServiceIntegrationTest.kt:160]: yes :thinking:

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 5]
```diff
@@ -106,56 +109,72 @@ internal class WorkspaceServiceImpl(
...
   ): Workspace {
     val existingWorkspace = this.getVerifiedWorkspace(organizationId, workspaceId, PERMISSION_WRITE)
     // Security cannot be changed by updateWorkspace
+    val updatedWorkspace = Workspace(
+      key = workspaceUpdateRequest.key ?: "",
+      name = workspaceUpdateRequest.name ?: "",
+      solution = workspaceUpdateRequest.solution ?: existingWorkspace.solution,
+      id = existingWorkspace.id,
+      organizationId = organizationId,
+      description = workspaceUpdateRequest.description ?: "",
```
[the remark this answers is not in the document]
Comment (sjoubert) [reply workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt]: Yeah I think this is broken, field will get emptied with the default `:? ""`. Maybe the organization update needs to be looked at again, not sure if the issue is there also.

As general design, all the PATCH method should probably follow a similar behavior. Maybe a [JSON merge patch](https://datatracker.ietf.org/doc/html/rfc7386) logic is applicable here. It can be very powerful, for example if the patch contains a `null`ed field, the semantic is to remove that field from the target object (if applicable).

Do not do it by hand, we need to find a way to do this, consistently, with the underlying storage objects.

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 6]
```diff
@@ -106,56 +109,77 @@ internal class WorkspaceServiceImpl(
...
+      name = workspaceUpdateRequest.name ?: existingWorkspace.name,
+      solution = workspaceUpdateRequest.solution ?: existingWorkspace.solution,
+      id = existingWorkspace.id,
+      organizationId = organizationId,
+      description = workspaceUpdateRequest.description ?: existingWorkspace.description,
+      tags = workspaceUpdateRequest.tags,
+      ownerId = existingWorkspace.ownerId,
+      webApp = workspaceUpdateRequest.webApp,
+      datasetCopy = workspaceUpdateRequest.datasetCopy,
+      security = existingWorkspace.security
```
Comment (jreynard-code) [inline workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:167]: Security should not be changed by users as we have all security endpoints associate to each resources

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 7]
```diff
@@ -106,56 +109,77 @@ internal class WorkspaceServiceImpl(
...
+    workspaceId: String,
+    workspaceUpdateRequest: WorkspaceUpdateRequest
   ): Workspace {
     val existingWorkspace = this.getVerifiedWorkspace(organizationId, workspaceId, PERMISSION_WRITE)
     // Security cannot be changed by updateWorkspace
+    val updatedWorkspace = Workspace(
+      key = workspaceUpdateRequest.key ?: existingWorkspace.key,
+      name = workspaceUpdateRequest.name ?: existingWorkspace.name,
+      solution = workspaceUpdateRequest.solution ?: existingWorkspace.solution,
+      id = existingWorkspace.id,
```
Comment (jreynard-code) [inline workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:160]: you can refer to `workspaceId` directly here

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 8]
```diff
@@ -106,56 +109,77 @@ internal class WorkspaceServiceImpl(
...
-      organizationId: String,
-      workspaceId: String,
-      workspace: Workspace
+    organizationId: String,
+    workspaceId: String,
+    workspaceUpdateRequest: WorkspaceUpdateRequest
   ): Workspace {
     val existingWorkspace = this.getVerifiedWorkspace(organizationId, workspaceId, PERMISSION_WRITE)
     // Security cannot be changed by updateWorkspace
+    val updatedWorkspace = Workspace(
```
Comment (jreynard-code) [inline workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:156]: It's still buggy here. 
All WorkspaceUpdateRequest optional fields should be tested.
If not tags/webApp/datasetCopy/... is set in `WorkspaceUpdateRequest`, all fields will be reset to null.
Please add test with empty `WorkspaceUpdateRequest` to cover this correctly

**Review by jreynard-code — CHANGES_REQUESTED** [review]
Some work to do but the end is close! Be strong ;)

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt (hunk 6 above)
Comment (Leopold-Cramer) [reply workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:167]: existingWorkspace.security is the already existing security, it's just needed to precise the security when creating a workspace as it is a required parameter

File: workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt [hunk 9]
```diff
@@ -557,43 +588,46 @@ internal class WorkspaceServiceImpl(
...
-      this.id,
-      this.security?.default ?: ROLE_NONE,
-      this.security?.accessControlList?.map { RbacAccessControl(it.id, it.role) }?.toMutableList()
-          ?: mutableListOf())
+  fun checkStringNotEmptyOrNull(values: List<String>){
+    val failingValues = mutableListOf<String>()
+    values.forEach { if(it.isNullOrEmpty()){failingValues.add(it)}}
+    if (failingValues.isNotEmpty())
+      throw IllegalArgumentException("${failingValues.toString()} is null or empty")
+  }
```
Comment (jreynard-code) [inline workspace/src/main/kotlin/com/cosmotech/workspace/service/WorkspaceServiceImpl.kt:619]: As discussed together, this can be removed

**Review by jreynard-code — CHANGES_REQUESTED** [review]
Some remarks.
Almost there ;)

**Review by jreynard-code — APPROVED** [review]
LGTM.
GG ;)

**Review by sjoubert — APPROVED** [review]
